Enhance e2e tests to toggle annotation on and off

girishkalele opened the pull request:
<!-- Reviewable:start -->

This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/31991)

<!-- Reviewable:end -->

bprashanth commented on test/e2e/service.go:
```diff
@@ -1099,6 +1099,89 @@ var _ = framework.KubeDescribe("Services", func() {
+			clientIP = content.String()
+			framework.Logf("ClientIP detected by target pod using VIP:SvcPort is %s", clientIP)
+			By("checking if Source IP is not preserved")
+			if strings.HasPrefix(clientIP, "10.") {
```
the more important thing to validate is that we actualy proxy across nodes. Someone turning this off is not going to care that src ip isn't preserved (in fact they probably would like it to be preserved). I'll take a todo. We can hit a node we know doesn't have the endpoint and make sure we get a response from another-node-endpoint instead.

girishkalele replied:
The problem with attempting to target a specific node is that we need to use destination address = VIP. The clientip is the only simple externally observable thing I can use in this e2e.

I actually want to add a new test that specifically places 2 endpoints on 1 node, and 1 endpoint on another, and then hit them enough times (using their /hostname endpoint) to get a statistical distribution across pods. With ESIPP off, we should see an even 33.33%, and with ESIPP on, we should see a 25/25/50% distribution.

For regular nodePorts, the logic has not been changed (yet).

bprashanth commented on test/e2e/service.go:
```diff
@@ -1099,6 +1099,89 @@ var _ = framework.KubeDescribe("Services", func() {
+				service.AnnotationValueExternalTrafficGlobal
+		})
+		success := false
+		for start := time.Now(); time.Since(start) < kubeProxyLagTimeout; time.Sleep(15 * time.Second) {
```
just use wait.PollImmediate? probably don't need success boolean in that case

bprashanth commented on test/e2e/service.go:
```diff
@@ -1099,6 +1099,89 @@ var _ = framework.KubeDescribe("Services", func() {
+			}
+		}
+		if !success {
+			framework.Logf("Source IP was preserved even when the ESIPP annotation was off")
```
why not fail?

bprashanth commented on test/e2e/service.go:
```diff
@@ -1099,6 +1099,89 @@ var _ = framework.KubeDescribe("Services", func() {
+		}
+		By("by deleting original service " + svc.Name)
+		err = jig.Client.Services(svc.Namespace).Delete(svc.Name)
+		Expect(err).NotTo(HaveOccurred())
```
there's a race here if this turns out to be the last test we run, and right after jenkins calls kube-down, we'll leak the loadbalancer and eventually our quota will build up. We should probably centralize Service deletion for this entire test in a function that not only deletes from apiserver, but also checks the type and polls till service controller has removed the resources from cloud. 

I'll take a todo since I suspect this entire suite runs into it.

girishkalele replied:
IIUC, you are saying that the service controller might not get to cleanup (because slow GCE API calls) and Jenkins kill kube-down and blow away all nodes before that happens ?

bprashanth commented on test/e2e/service.go:
```diff
@@ -1099,6 +1099,89 @@ var _ = framework.KubeDescribe("Services", func() {
+			By("reading clientIP using the TCP service's service port via its external VIP")
+			content = jig.GetHTTPContent(ingressIP, svcTcpPort, kubeProxyLagTimeout, "/clientip")
+			clientIP = content.String()
+			framework.Logf("ClientIP detected by target pod using VIP:SvcPort is %s", clientIP)
```
can you only log on failure? like ip Failf("returned from service different from vip"), or something? All this is noise in test logs if the failure is elsewhere in this test. Ideally we'd just see something like:
```
Step: changing ESIPP annotation off and checking cross node proxy
Step: changing ESIPP back on and checking source ip
Step: something else
```

and when we see a failure: 
```
Step: one
Step: two
  Failed to verify source ip, VIP returned blah service returned Blah
Step: three
```

I'll take a follow up if you don't want to do it in this one, but this test is already gigantic.

girishkalele replied:
Is there a lower log level that will only show up with some super verbose ginkgo log level ?
That way, whenever I need to debug this test, I can turn on these logs.

girishkalele replied:
How about writing to a byte.Buffer and only printing out the current buffer when there is a Failf ?

Also, I see there is an unexported log() function in test/e2e/framework/util.go (line 348), but it doesn't have a log level filter check.

```
func log(level string, format string, args ...interface{}) {
	fmt.Fprintf(GinkgoWriter, nowStamp()+": "+level+": "+format+"\n", args...)
}

func Logf(format string, args ...interface{}) {
	log("INFO", format, args...)
}
```

girishkalele commented:
Review status: 0 of 1 files reviewed at latest revision, 5 unresolved discussions.

---

*[test/e2e/service.go, line 1110 \[r1\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31991#-KQg-QuJBSsWF3_0ovkp-r1-1110:-KR9s5L9CacaIAimTr0-:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/563d36f2717542bb925a5e498e3996f4a0c7bdba/test/e2e/service.go#L1110)):*
<details><summary><i>Previously, bprashanth (Prashanth B) wrote…</i></summary>
> just use wait.PollImmediate? probably don't need success boolean in that case
</details>
Done.

---

*[test/e2e/service.go, line 1118 \[r1\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31991#-KQg-QuJBSsWF3_0ovkp-r1-1118:-KR9s88l4HT9hR9BFil1:bly79wl) ([raw file](https://github.com/kubernetes/kubernetes/blob/563d36f2717542bb925a5e498e3996f4a0c7bdba/test/e2e/service.go#L1118)):*
<details><summary><i>Previously, girishkalele (Girish Kalele) wrote…</i></summary>
> How about writing to a byte.Buffer and only printing out the current buffer when there is a Failf ?
> 
> 
> 
> Also, I see there is an unexported log() function in test/e2e/framework/util.go (line 348), but it doesn't have a log level filter check.
> 
> 
> 
> ```
> 
> func log(level string, format string, args ...interface{}) {
> 
> 	fmt.Fprintf(GinkgoWriter, nowStamp()+": "+level+": "+format+"\n", args...)
> 
> }
> 
> 
> 
> func Logf(format string, args ...interface{}) {
> 
> 	log("INFO", format, args...)
> 
> }
> 
> ```
</details>
Switched the serviceTestJig to buffer all Logf messages and only print them when there is a Failf message being printed.

---

*[test/e2e/service.go, line 1127 \[r1\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31991#-KQg-QuJBSsWF3_0ovkp-r1-1127:-KR9sHei1DIqsG0Y2kAq:baps5wc) ([raw file](https://github.com/kubernetes/kubernetes/blob/563d36f2717542bb925a5e498e3996f4a0c7bdba/test/e2e/service.go#L1127)):*
<details><summary><i>Previously, bprashanth (Prashanth B) wrote…</i></summary>
> why not fail? 
</details>
Changed to Fail.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31991)*
<!-- Sent from Reviewable.io -->

thockin commented:
ping @bprashanth

bprashanth commented on test/e2e/service.go:
```diff
@@ -1145,12 +1145,93 @@ var _ = framework.KubeDescribe("Services", func() {
+			jig.Failf("Service HealthCheck NodePort annotation not present")
+		}
+		By("by deleting original service " + svc.Name)
+		err = jig.Client.Services(svc.Namespace).Delete(svc.Name)
```
you need to wait on this service or wait for Jessies pr which polls in aftereach, so we avoid resource leak

bprashanth commented on test/e2e/service.go:
```diff
@@ -1145,12 +1145,93 @@ var _ = framework.KubeDescribe("Services", func() {
+			// Request the same healthCheckNodePort as before, to test the user-requested allocation path
+			svc.ObjectMeta.Annotations[service.AnnotationHealthCheckNodePort] =
+				fmt.Sprintf("%d", healthCheckNodePort)
+		})
```
anyway you can actually refactor into a function? you're basically checking for the same invariant before/after toggling the annotations: 
```
checkESIPPInvariants(preserved=true,false) {
  gethttpcontent
  clientip = content.string()
  if preserved && !HasPrefix("10.") {
    failf()
  } else if HasPrefix() {
    ...
  }
}
```

If there's a bug in the logic I'd really like to fix one place. It's also clearer what one expects with the annotation.
But if the refactor is too messy, never mind.

bprashanth commented on test/e2e/service.go:
```diff
@@ -1145,12 +1145,93 @@ var _ = framework.KubeDescribe("Services", func() {
+			jig.Logf("Checking health check response for node %s, public IP %s", nodes.Items[n].Name, publicIP)
+			// All nodes should fail the healthcheck on the service healthCheckNodePort
+			jig.TestHTTPHealthCheckNodePort(publicIP, healthCheckNodePort, "/healthz", false)
+		}
```
it's worth checking, if we can, that the previously allocated nodeport is free by creating a test servie with that exact nodeport. Risk is that we need to ensure this test is run serially. It's already `[Slow]`, not sure if all Slow tests are also serial. 

Whatever you decide, please add a todo.

bprashanth commented on test/e2e/service.go:
```diff
@@ -1145,12 +1145,93 @@ var _ = framework.KubeDescribe("Services", func() {
 			// HealthCheck should pass only on the node where num(endpoints) > 0
 			// All other nodes should fail the healthcheck on the service healthCheckNodePort
 			expectedSuccess := nodes.Items[n].Name == readyHostName
 			jig.TestHTTPHealthCheckNodePort(publicIP, healthCheckNodePort, "/healthz", expectedSuccess)
```
this request should in theory just get rst and drop immediately, but if the remote end fails to send rst, does it just hang or timeout client side? if so, what's the client timeout?

bprashanth commented on test/e2e/service.go:
```diff
@@ -1117,25 +1117,25 @@ var _ = framework.KubeDescribe("Services", func() {
-			framework.Failf("Expected Ready endpoints - found none")
+			jig.Failf("Expected Ready endpoints - found none")
 		}
 		readyHostName := *endpoints.Subsets[0].Addresses[0].NodeName
```
readyHostName is confusing, makes me think it's the name of a ready node but it's really the name of a node a ready endpoint is on. Suggest endpointNodeName.

bprashanth commented:
mostly nits

bprashanth commented:
to summarize:     

 * whats the client timeout for http request
 * check for nodeport leaks
 * request for refactor

You can punt on the last 2 with todos

girishkalele commented:
Working on them, will ping when ready.

girishkalele commented:
Review status: 0 of 1 files reviewed at latest revision, 10 unresolved discussions.

---

*[test/e2e/service.go, line 1137 \[r2\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31991#-KQg-QuJBSsWF3_0ovkp-r2-1137:-KR_Z9N8orgWel_Lo9bf:b-a6gx7v) ([raw file](https://github.com/kubernetes/kubernetes/blob/eb7a02adc5122504de2b33a744e474c161d92b08/test/e2e/service.go#L1137)):*
<details><summary><i>Previously, bprashanth (Prashanth B) wrote…</i></summary>
> readyHostName is confusing, makes me think it's the name of a ready node but it's really the name of a node a ready endpoint is on. Suggest endpointNodeName. 
</details>
Will change.

---

*[test/e2e/service.go, line 1152 \[r2\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31991#-KQg-QuJBSsWF3_0ovkp-r2-1152:-KR_Y46wI080ZEI_NkyH:b-6c7u60) ([raw file](https://github.com/kubernetes/kubernetes/blob/eb7a02adc5122504de2b33a744e474c161d92b08/test/e2e/service.go#L1152)):*
<details><summary><i>Previously, bprashanth (Prashanth B) wrote…</i></summary>
> this request should in theory just get rst and drop immediately, but if the remote end fails to send rst, does it just hang or timeout client side? if so, what's the client timeout? 
</details>
This is calling httpGetNoConnectionPool which internally has a hardcoded 5 * time.Seconds timeout for the http.Client created.
If the nodeport is not completely healthy or iptables drops the incoming SYN, hopefully, this means the http.Client.Get() will timeout and return failure in 5 seconds.
Added a comment with the 5 second timeout in the function comment.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31991)*
<!-- Sent from Reviewable.io -->

girishkalele commented:
Review status: 0 of 1 files reviewed at latest revision, 10 unresolved discussions.

---

*[test/e2e/service.go, line 1189 \[r2\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31991#-KQg-QuJBSsWF3_0ovkp-r2-1189:-KR__DxgYgZV1Am5UwZn:bqwwcaf) ([raw file](https://github.com/kubernetes/kubernetes/blob/eb7a02adc5122504de2b33a744e474c161d92b08/test/e2e/service.go#L1189)):*
<details><summary><i>Previously, bprashanth (Prashanth B) wrote…</i></summary>
> it's worth checking, if we can, that the previously allocated nodeport is free by creating a test servie with that exact nodeport. Risk is that we need to ensure this test is run serially. It's already `[Slow]`, not sure if all Slow tests are also serial. 
> 
> 
> 
> Whatever you decide, please add a todo. 
</details>
I will add a TODO to note that we need to add this verification.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31991)*
<!-- Sent from Reviewable.io -->

bprashanth commented:
*[test/e2e/service.go, line 1183 \[r1\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31991#-KQg-QuJBSsWF3_0ovkp-r1-1183:-KR__MA579nklmsuNl4o:b2nign6) ([raw file](https://github.com/kubernetes/kubernetes/blob/563d36f2717542bb925a5e498e3996f4a0c7bdba/test/e2e/service.go#L1183)):*
<details><summary><i>Previously, girishkalele (Girish Kalele) wrote…</i></summary>
> IIUC, you are saying that the service controller might not get to cleanup (because slow GCE API calls) and Jenkins kill kube-down and blow away all nodes before that happens ?
</details>
yes

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31991)*
<!-- Sent from Reviewable.io -->

bprashanth commented:
*[test/e2e/service.go, line 1120 \[r1\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31991#-KQg-QuJBSsWF3_0ovkp-r1-1120:-KR__UE9amKnpCEvqcUL:b-1jx56m) ([raw file](https://github.com/kubernetes/kubernetes/blob/563d36f2717542bb925a5e498e3996f4a0c7bdba/test/e2e/service.go#L1120)):*
<details><summary><i>Previously, girishkalele (Girish Kalele) wrote…</i></summary>
> The problem with attempting to target a specific node is that we need to use destination address = VIP. The clientip is the only simple externally observable thing I can use in this e2e.
> 
> 
> 
> I actually want to add a new test that specifically places 2 endpoints on 1 node, and 1 endpoint on another, and then hit them enough times (using their /hostname endpoint) to get a statistical distribution across pods. With ESIPP off, we should see an even 33.33%, and with ESIPP on, we should see a 25/25/50% distribution.
> 
> 
> 
> For regular nodePorts, the logic has not been changed (yet).
</details>
I swear I filed a bug for that

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31991)*
<!-- Sent from Reviewable.io -->

girishkalele commented:
Review status: 0 of 1 files reviewed at latest revision, 4 unresolved discussions.

---

*[test/e2e/service.go, line 1232 \[r2\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31991#-KQg-QuJBSsWF3_0ovkp-r2-1232:-KR_aPt39kWobMm1YMgs:bmrkmkd) ([raw file](https://github.com/kubernetes/kubernetes/blob/eb7a02adc5122504de2b33a744e474c161d92b08/test/e2e/service.go#L1232)):*
<details><summary><i>Previously, bprashanth (Prashanth B) wrote…</i></summary>
> you need to wait on this service or wait for Jessies pr which polls in aftereach, so we avoid resource leak
</details>
I think Jesse's AfterEach fix will be going in soon, don't want to duplicate that functionality here.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31991)*
<!-- Sent from Reviewable.io -->

Commit:
Enhance e2e tests to toggle annotation on and off

girishkalele commented:
@k8s-bot test this ignore #33184

girishkalele commented:
@bprashanth Ready to review.
Implemented #1, the code factoring looks messy so will punt the other two items to follow-on PRs.
@mfanjie has PR #32183 in-flight, that should take care of cleaning up as your suggested at the end of the e2e test.
